File: solutions/javascript/isbn-verifier/1/isbn-verifier.js


export function isValid(isbn) {

  // Validate applies the formula to the ISBN 9 digits and bonus digit.
  const validate = function([x1, x2, x3, x4, x5, x6, x7, x8, x9, x10]) {
    const formula = (x1 * 10 + x2 * 9 + x3 * 8 + x4 * 7 + x5 * 6 + x6 * 5 + x7 * 4 + x8 * 3 + x9 * 2 + x10 * 1);
    return (formula % 11 == 0);
  }

  // parseDigit parse the digits string (0 to 9 and X) into numbers.
  const parseDigit = function(d) {
    if (d == 'X') return 10;
    return Number.parseInt(d);
  } 

  // Remove the hyphens, check the number is 9 digits followed by a digit or X
  // and then check it.
  let cleanedUpIsbn = isbn.replace(/-/g, '');
  if (! /^[0-9]{9}[0-9X]$/.test(cleanedUpIsbn)) return false;

  return validate(cleanedUpIsbn.split('').map(d => parseDigit(d)));
  
};

File: solutions/lua/isbn-verifier/1/isbn-verifier.lua
local function isbnToDigits(isbn)

  local res = {}
  for char in isbn:gmatch('.') do
    local digit = (char == 'X') and 10 or tonumber(char)
    table.insert(res, digit)
  end

  if #res == 9 then
    table.insert(res, 0)
  end

  return res
end

local function validate(x)

  local formula = x[1] * 10 + x[2] * 9 + x[3] * 8 + x[4] * 7 + x[5] * 6 + x[6] * 5
      + x[7] * 4 + x[8] * 3 + x[9] * 2 + x[10] * 1
  return formula % 11 == 0

end

return {
  valid = function(isbn)

    -- remove hyphens
    local cleaned_up_isbn = isbn:gsub('-', '')

    -- check the isbn is valid
    if not cleaned_up_isbn:match('^%d%d%d%d%d%d%d%d%d[0-9X]?$') then return false end

    return validate(isbnToDigits(cleaned_up_isbn))

  end
}
